ultralytics/engine/trainer.py

```diff
@@ -300,10 +315,23 @@ def _setup_train(self):
         if RANK > -1 and self.world_size > 1:  # DDP
             dist.broadcast(self.amp.int(), src=0)  # broadcast from rank 0 to all other ranks; gloo errors with boolean
         self.amp = bool(self.amp)  # as boolean
-        self.scaler = (
-            torch.amp.GradScaler("cuda", enabled=self.amp) if TORCH_2_4 else torch.cuda.amp.GradScaler(enabled=self.amp)
-        )
-        if self.world_size > 1:
+
+        if self.device.type == "cuda":
+            self.scaler = (
+                torch.amp.GradScaler("cuda", enabled=self.amp)
+                if TORCH_2_4
+                else torch.cuda.amp.GradScaler(enabled=self.amp)
+            )
+        elif self.device.type == "xpu":
+            self.scaler = torch.amp.GradScaler(self.device, enabled=self.amp)
+        else:
+            # Handle CPU, MPS and other devices
+            if TORCH_2_4:
+                self.scaler = torch.amp.GradScaler(device="cpu", enabled=self.amp)
+            else:
+                self.scaler = torch.cuda.amp.GradScaler(enabled=self.amp)
+
+        if world_size > 1:
```

Comment:
❗ **CRITICAL**: `world_size` is undefined here, so this will crash as soon as we hit the DDP branch. Please keep using `self.world_size`.

**Suggested change:**
```suggestion
        if self.world_size > 1:
```